### PR TITLE
${exception) - Improved logging AggregrateException's Data

### PR DESCRIPTION
**Pull request by chaos0307**
Current version of NLog is not able to log AggregrateException.Data as it will be lost after it was Flatten. Fixed it by assigned it to a temporary variable and add it to the newly flatten exception.

Unit test is include as well. 

Cheers.

### CODE REVIEW

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 1]
```diff
@@ -229,8 +230,12 @@ protected override void Append(StringBuilder builder, LogEventInfo logEvent)
 #if !NET3_5 && !SILVERLIGHT4
                 if (logEvent.Exception is AggregateException aggregateException)
                 {
+                    var aggregateExceptionData = aggregateException.Data;
                     aggregateException = aggregateException.Flatten();
                     primaryException = GetPrimaryException(aggregateException);
+                    foreach (var key in aggregateExceptionData.Keys)
+                        primaryException.Data.Add(key, aggregateExceptionData[key]);
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs:237]: Prefer that the logging-library only probes the Exception, and doesn't modify the Exceptions.

**Comment by snakefoot** [comment]
Maybe something like this:

```c#
        private void AppendException(Exception currentException, List<ExceptionRenderingFormat> renderFormats, StringBuilder builder, Exception aggregateException)
        {
            int currentLength = builder.Length;
            foreach (ExceptionRenderingFormat renderingFormat in renderFormats)
            {
                int beforeRenderLength = builder.Length;
                var currentRenderFunction = _renderingfunctions[renderingFormat];
                if (renderingFormat == ExceptionRenderingFormat.Data)
                {
                     var dataException = (aggregateException?.Data?.Count > 0 && currentException?.Data?.Count == 0) ? aggregateException : currentException;
                     currentRenderFunction(builder, dataException);
                }
                else
                {
                   currentRenderFunction(builder, currentException);
                }
                if (builder.Length != beforeRenderLength)
                {
                    currentLength = builder.Length;
                    builder.Append(Separator);
                }
            }

            builder.Length = currentLength;
        }
```

Alternative one should perform the lookup of `primaryException` inside each ExceptionRenderingFormat-function. And then the `AppendData` could have the exotic logic of scanning the entire exception-tree.

File: src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs [hunk 2]
```diff
@@ -309,15 +305,27 @@ private void AppendInnerException(Exception currentException, StringBuilder buil
             builder.Append(InnerExceptionSeparator);
             AppendException(currentException, InnerFormats ?? Formats, builder);
         }
-
+#if !NET3_5 && !SILVERLIGHT4
+        private void AppendException(Exception currentException, List<ExceptionRenderingFormat> renderFormats, StringBuilder builder, AggregateException aggregateException = null)
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs:309]: If you change it to `Exception aggregateException = null`, then you can remove `#if !NET3_5 && !SILVERLIGHT4`

Comment (chaos0307) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs:309]: i think i'll keep the `#if` and input param as type AggregateException as it only available post NET3_5

Comment (304NotModified) [reply src/NLog/LayoutRenderers/ExceptionLayoutRenderer.cs:309]: I prefer no `#if` if possible. As that is hard with refactoring. 👼 (no refactoring tool will handle those fully correct when moving)

**Comment by snakefoot** [comment]
Happy that the `#if` is now gone. Could one also remove the if-statement. So it becomes:

```c#
_renderingfunctions = new Dictionary<ExceptionRenderingFormat, Action<StringBuilder, Exception,Exception>>()
{
{ExceptionRenderingFormat.Message, (sb, ex, aggex) => AppendMessage(sb, ex)},
{ExceptionRenderingFormat.Type, (sb, ex, aggex) => AppendType(sb, ex)},
{ExceptionRenderingFormat.ShortType, (sb, ex, aggex) => AppendShortType(sb, ex)},
{ExceptionRenderingFormat.ToString, (sb, ex, aggex) => AppendToString(sb, ex)},
{ExceptionRenderingFormat.Method, (sb, ex, aggex) => AppendMethod(sb, ex)},
{ExceptionRenderingFormat.Source, (sb, ex, aggex) => AppendSource(sb, ex)},
{ExceptionRenderingFormat.StackTrace, (sb, ex, aggex) => AppendStackTrace(sb, ex)},
{ExceptionRenderingFormat.Data, (sb, ex, aggex) => AppendData(sb, ex, aggex)},
{ExceptionRenderingFormat.Serialize, (sb, ex, aggex) => AppendSerializeObject(sb, ex)},
{ExceptionRenderingFormat.HResult, (sb, ex, aggex) => AppendHResult(sb, ex)},
{ExceptionRenderingFormat.Properties, (sb, ex, aggex) => AppendProperties(sb, ex)},
};
```

And then have this:

```c#
        private void AppendException(Exception currentException, List<ExceptionRenderingFormat> renderFormats, StringBuilder builder, Exception aggregateException = null)
        {
            int currentLength = builder.Length;
            foreach (ExceptionRenderingFormat renderingFormat in renderFormats)
            {
                int beforeRenderLength = builder.Length;
                var currentRenderFunction = _renderingfunctions[renderingFormat];
                currentRenderFunction(builder, currentException, aggregateException );
                if (builder.Length != beforeRenderLength)
                {
                    currentLength = builder.Length;
                    builder.Append(Separator);
                }
            }

            builder.Length = currentLength;
        }
```

And then have this new method:

```c#
private void AppendData(StringBuilder builder, Exception ex, Exception aggregateException = null)
{
      if (aggregateException?.Data?.Count > 0)
      {
            AppendData(builder, aggregateException);
            builder.Append(Separator);
      }
      AppendData(builder, ex);
}
```

**Comment by snakefoot** [comment]
Could also consider to get rid of the Flatten-logic by default:

```c#
private static Exception GetPrimaryException(AggregateException aggregateException)
{
   if (aggregateException.InnerExceptions.Count == 1)
   {
       var innerException = aggregateException.InnerExceptions[0];
       if (!(innerException is AggregateException))
            return innerException;  // Skip calling Flatten()
   }
 
   aggregateException = aggregateException.Flatten();
   if (aggregateException.InnerExceptions.Count == 1)
   {
       return aggregateException.InnerExceptions[0];
   }
 
   return aggregateException;
}
```

```c#
#if !NET3_5 && !SILVERLIGHT4
                if (logEvent.Exception is AggregateException aggregateException)
                {
                   primaryException = GetPrimaryException(aggregateException);
                   AppendException(primaryException, Formats, builder, aggregateException);
                   if (currentLevel < MaxInnerExceptionLevel)
                   {
                      currentLevel = AppendInnerExceptionTree(primaryException, currentLevel, builder);
                      if (currentLevel < MaxInnerExceptionLevel)
                      {
                          aggregateException = primaryException as AggregateException;
                          if (aggregateException?.InnerExceptions?.Count > 1)
                          {
                             AppendAggregateException(aggregateException, currentLevel, builder);
                          }
                      }
                   }
                }
```

**Comment by snakefoot** [comment]
Could consider to rebase to "master", then it will not wait for NLog 5.0

**Comment by snakefoot** [comment]
@chaos0307 Wow you are fast. Then just need to make the [code-smell-engine](https://sonarcloud.io/project/issues?id=nlog2&pullRequest=3974&resolved=false&types=CODE_SMELL) happy (by removing default parameter value `= null`)

`private void AppendData(StringBuilder builder, Exception ex, Exception aggregateException = null)`

> This method signature overlaps the one defined on line 467, the default parameter value can only be used with named arguments.Why is this an issue?

**Comment by 304NotModified** [comment]
Cool! I think this is done now, agree @snakefoot ?

**Comment by snakefoot** [comment]
@304NotModified Yes I agree. Now just need to decide if for NLog 4.7.3 (`master`) or for NLog 5.0 (`dev`)

**Comment by 304NotModified** [comment]
I think 4.7.3 is better. Agree?

If so, @chaos0307 could you please rebase this on master? (maybe squash first). If you need help, let us know!

**Comment by snakefoot** [comment]
Also think NLog 4.7.3 is good.

**Comment by chaos0307** [comment]
@304NotModified should i rebase to dev branch and then to master branch? the current working branch is branch out from dev though

**Comment by 304NotModified** [comment]
I've squased and rebased on master :)

**Comment by 304NotModified** [comment]
(PS sonarcloud has issues, outside our control)

**Pull request merged**